libs/framework/src/bundle_archive.c

```diff
@@ -28,644 +28,451 @@
+ * The lifecycle of a bundle archive is coupled to the lifecycle of the bundle that is created from the archive.
+ *
+ * @note The bundle archive is thread safe.
+ */
 struct bundleArchive {
+    //initialed during creation and immutable
 	celix_framework_t* fw;
 	long id;
-	char * location;
-	DIR *archiveRootDir;
-	char * archiveRoot;
-	linked_list_pt revisions;
-	long refreshCount;
-	time_t lastModified;
-
-	bundle_state_e persistentState;
-};
+	char *archiveRoot;
+    char *savedBundleStatePropertiesPath;
+    char* storeRoot;
+    bool isSystemBundle;
+    char* bundleSymbolicName; //read from the manifest
+    char* bundleVersion; //read from the manifest
 
-static celix_status_t bundleArchive_getRevisionLocation(bundle_archive_pt archive, long revNr, char **revision_location);
-static celix_status_t bundleArchive_setRevisionLocation(bundle_archive_pt archive, const char * location, long revNr);
 
-static celix_status_t bundleArchive_initialize(bundle_archive_pt archive);
+    celix_thread_mutex_t lock; //protects below and saving of bundle state properties
 
-static celix_status_t bundleArchive_createRevisionFromLocation(bundle_archive_pt archive, const char *location, const char *inputFile, long revNr, bundle_revision_pt *bundle_revision);
-static celix_status_t bundleArchive_reviseInternal(bundle_archive_pt archive, bool isReload, long revNr, const char * location, const char *inputFile);
+    char* location;
+    char* currentRevisionRoot;
+    celix_array_list_t* revisions; //list of bundle_revision_t*
 
-static celix_status_t bundleArchive_readLastModified(bundle_archive_pt archive, time_t *time);
-static celix_status_t bundleArchive_writeLastModified(bundle_archive_pt archive);
+    //bundle cache state info
+	long revisionNr;
 
-celix_status_t bundleArchive_createSystemBundleArchive(celix_framework_t* fw, bundle_archive_pt *bundle_archive) {
-	celix_status_t status = CELIX_SUCCESS;
-	char *error = NULL;
-	bundle_archive_pt archive = NULL;
-
-	if (*bundle_archive != NULL) {
-		status = CELIX_ILLEGAL_ARGUMENT;
-		error = "Missing required arguments and/or incorrect values";
-	} else {
-		archive = (bundle_archive_pt) calloc(1,sizeof(*archive));
-		if (archive == NULL) {
-			status = CELIX_ENOMEM;
-		} else {
-			status = linkedList_create(&archive->revisions);
-			if (status == CELIX_SUCCESS) {
-				archive->fw = fw;
-				archive->id = CELIX_FRAMEWORK_BUNDLE_ID;
-				archive->location = strdup("System Bundle");
-				archive->archiveRoot = NULL;
-				archive->archiveRootDir = NULL;
-				archive->refreshCount = -1;
-				archive->persistentState = CELIX_BUNDLE_STATE_UNKNOWN;
-				time(&archive->lastModified);
-
-				*bundle_archive = archive;
-			}
-		}
-	}
-
-	if(status != CELIX_SUCCESS && archive != NULL){
-		bundleArchive_destroy(archive);
-	}
-
-	framework_logIfError(fw->logger, status, error, "Could not create archive");
+    //bundle cache state in properties form
+    celix_properties_t* bundleStateProperties;
```

Comment:
I notice that it's only used in `celix_bundleArchive_updateAndStoreBundleStateProperties`.
Why keep it in memory?

Reply:
This was based on the previous bundle cache implementation. 
The idea behind that implementation was also to restart from existing bundle caches. 

This could be interresting for the future, but for now I want to focus on a bundle cache that can be configured to only extract zip files if needed. I will remove the bundleStateProperties from the bundle archive object.